nemo/collections/nlp/data/language_modeling/megatron/dataset_utils.py

```diff
@@ -38,7 +38,16 @@
 
 import numpy as np
 import torch
-from apex.transformer import parallel_state
+
+try:
+    from apex.transformer import parallel_state
+
+    HAVE_APEX = True
+
+except (ImportError, ModuleNotFoundError):
+
+    HAVE_APEX = False
```

Comment:
Just a note, no need for so many new lines

Reply:
Also, anything that calls apex.*  Needs to first read this value of Have_APEX and then only use parallel_state. Otherwise if will fail with an import error (say in a class init)

Reply:
This is insane we can't wrap every call to apex in an if statement. If they don't have apex, installed they'll get an appropriate error from python.

Reply:
Then you'll still get errors when your import failed but some class calls some function which depends on apex. Sure, you can leave if out for now but it will have issues downstream

Reply:
Btw all the optional code does exactly this - check the import and then raise appropriate error for every single call in every single file. It is nothing special for apex but it's upto the NMT team to decide if they are ok with partial checks like this.

Reply:
Well python will raise an error anyway - so it's fine. But if there are dependent calls mixed into places where apex is actually optional, then the code may fail again. For now it's fine